`quarkus.test.integration-test-profile` ignored in IT startup error message if startup fails because of ConfigValidationException

### Describe the bug

If I set this:

```properties
quarkus.test.integration-test-profile=integrationtest
```

and for some reason my application fails to start in integration tests, **during config resolution**, the error message looks like this:

```
ERROR: Failed to start application (with profile [prod])
java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
```

... even though the profile prod is NOT active, as evidenced by the command being executed:

```
Executing "... -Dquarkus.profile=integrationtest ..."
```

### Expected behavior

```
ERROR: Failed to start application (with profile [integrationtest])
```

### Actual behavior

```
ERROR: Failed to start application (with profile [prod])
java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:111)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: io.smallrye.config.ConfigValidationException: Configuration validation failed:
        java.util.NoSuchElementException: SRCFG00011: Could not expand value maven.test.data.path in property uris.quarkusio.local
        at io.smallrye.config.ConfigMappingProvider.mapConfigurationInternal(ConfigMappingProvider.java:1001)
        at io.smallrye.config.ConfigMappingProvider.lambda$mapConfiguration$3(ConfigMappingProvider.java:961)
        at io.smallrye.config.SecretKeys.doUnlocked(SecretKeys.java:28)
        at io.smallrye.config.ConfigMappingProvider.mapConfiguration(ConfigMappingProvider.java:961)
        at io.smallrye.config.ConfigMappings.mapConfiguration(ConfigMappings.java:91)
        at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:647)
        at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
        at io.quarkus.runtime.generated.Config.createRunTimeConfig(Unknown Source)
        at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
        ... 7 more
```

### How to Reproduce?

Just adding something like this to an application should have done the trick:

```properties
quarkus.test.integration-test-profile=integrationtest
%integrationtest.hello.message=${some-unresolved-property}
```

Assuming the application also includes this:

```java
@Path("/hello")
public class GreetingResource {

    @ConfigMapping(prefix = "hello")
    public interface MyConfig {
        String message();
    }

    @Inject
    MyConfig config;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return config.message();
    }
}
```

But here's a full reproducer:

```shell
git clone -b i36376 https://github.com/yrodiere/quarkus-playground.git
cd quarkus-playground
./mvnw clean verify -Dnative
```

When build ends (it'll take a while), see how the logs are inconsistent:

```
[...]
Executing "/home/yrodiere/workspaces/testcases/quarkus-playground/target/code-with-quarkus-1.0.0-SNAPSHOT-runner -Dquarkus.http.port=8081 -Dquarkus.http.ssl-port=8444 -Dtest.url=http://localhost:8081 -Dquarkus.log.file.path=/home/yrodiere/workspaces/testcases/quarkus-playground/target/quarkus.log -Dquarkus.log.file.enable=true -Dquarkus.log.category."io.quarkus".level=INFO -Dquarkus.profile=integrationtest -Dquarkus.datasource.password=quarkus -Dquarkus.datasource.db-kind=postgresql -Dquarkus.datasource.jdbc.url=jdbc:postgresql://localhost:43893/quarkus?loggerLevel=OFF -Dquarkus.datasource.username=quarkus -Dquarkus.hibernate-orm.database.generation=drop-and-create"
[...]
ERROR: Failed to start application (with profile [prod])
java.lang.RuntimeException: Failed to start quarkus
[...]
```

The first line shows `-Dquarkus.profile=integrationtest` but the error shows `with profile [prod]`.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

I think this can only be reproduced if the startup failure happens during config resolution. That would make sense, as in that case I suspect `ConfigUtils.getProfiles()` just returns garbage that doesn't take config into account, so this would not behave properly:

https://github.com/quarkusio/quarkus/blob/700e37b6dca14771c771dc969b66e41e7c6684e0/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java#L196-L199

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/RuntimeConfigSetupBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/RuntimeConfigSetupBuildStep.java
@@ -10,11 +10,14 @@ import io.quarkus.deployment.annotations.Produce;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.MainBytecodeRecorderBuildItem;
 import io.quarkus.deployment.builditem.RuntimeConfigSetupCompleteBuildItem;
+import io.quarkus.gizmo.CatchBlockCreator;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.ClassOutput;
 import io.quarkus.gizmo.MethodCreator;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.runtime.StartupContext;
 import io.quarkus.runtime.StartupTask;
+import io.quarkus.runtime.configuration.ConfigurationException;
 
 public class RuntimeConfigSetupBuildStep {
     private static final String RUNTIME_CONFIG_STARTUP_TASK_CLASS_NAME = "io.quarkus.deployment.steps.RuntimeConfigSetup";
@@ -37,11 +40,17 @@ public class RuntimeConfigSetupBuildStep {
                 .interfaces(StartupTask.class).build()) {
 
             try (MethodCreator method = clazz.getMethodCreator("deploy", void.class, StartupContext.class)) {
-                method.invokeVirtualMethod(ofMethod(StartupContext.class, "setCurrentBuildStepName", void.class, String.class),
+                TryBlock tryBlock = method.tryBlock();
+                tryBlock.invokeVirtualMethod(
+                        ofMethod(StartupContext.class, "setCurrentBuildStepName", void.class, String.class),
                         method.getMethodParam(0), method.load("RuntimeConfigSetupBuildStep.setupRuntimeConfig"));
 
-                method.invokeStaticMethod(C_CREATE_RUN_TIME_CONFIG);
-                method.returnValue(null);
+                tryBlock.invokeStaticMethod(C_CREATE_RUN_TIME_CONFIG);
+                tryBlock.returnValue(null);
+
+                CatchBlockCreator cb = tryBlock.addCatch(RuntimeException.class);
+                cb.throwException(ConfigurationException.class, "Failed to read configuration properties",
+                        cb.getCaughtException());
             }
         }
 

--- a/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
@@ -187,8 +187,9 @@ public class ApplicationLifecycleManager {
                         }
                         applicationLogger.warn("You can try to kill it with 'kill -9 <pid>'.");
                     }
-                } else if (rootCause instanceof ConfigurationException) {
+                } else if (ExceptionUtil.isAnyCauseInstanceOf(e, ConfigurationException.class)) {
                     System.err.println(rootCause.getMessage());
+                    e.printStackTrace();
                 } else if (rootCause instanceof PreventFurtherStepsException
                         && !StringUtil.isNullOrEmpty(rootCause.getMessage())) {
                     System.err.println(rootCause.getMessage());

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ExceptionUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ExceptionUtil.java
@@ -86,6 +86,17 @@ public class ExceptionUtil {
         return chain.isEmpty() ? null : chain.get(chain.size() - 1);
     }
 
+    public static <T> boolean isAnyCauseInstanceOf(Throwable exception, Class<T> classToCheck) {
+        Throwable curr = exception;
+        do {
+            if (classToCheck.isInstance(curr)) {
+                return true;
+            }
+            curr = curr.getCause();
+        } while (curr != null);
+        return false;
+    }
+
     /**
      * Creates and returns a new {@link Throwable} which has the following characteristics:
      * <ul>

--- a/core/runtime/src/test/java/io/quarkus/runtime/util/ExceptionUtilTest.java
+++ b/core/runtime/src/test/java/io/quarkus/runtime/util/ExceptionUtilTest.java
@@ -13,6 +13,8 @@ import java.util.List;
 
 import org.junit.jupiter.api.Test;
 
+import io.quarkus.runtime.configuration.ConfigurationException;
+
 /**
  *
  */
@@ -59,6 +61,11 @@ public class ExceptionUtilTest {
         assertEquals(NullPointerException.class, ExceptionUtil.getRootCause(new NullPointerException()).getClass());
     }
 
+    @Test
+    public void testIsAnyCauseInstanceOf() {
+        assertTrue(ExceptionUtil.isAnyCauseInstanceOf(generateConfigurationException(), ConfigurationException.class));
+    }
+
     private Throwable generateException() {
         try {
             try {
@@ -79,4 +86,26 @@ public class ExceptionUtilTest {
         }
         throw new RuntimeException("Should not reach here");
     }
+
+    private Throwable generateConfigurationException() {
+        try {
+            try {
+                Integer.parseInt("23.23232");
+            } catch (NumberFormatException nfe) {
+                throw new ConfigurationException("Incorrect param", nfe);
+            }
+        } catch (ConfigurationException ce) {
+            try {
+                throw new IOException("Request processing failed", ce);
+            } catch (IOException e) {
+                try {
+                    throw new IOError(e);
+                } catch (IOError ie) {
+                    return new RuntimeException("Unexpected exception", ie);
+                }
+            }
+        }
+        throw new RuntimeException("Should not reach here");
+    }
+
 }

